File: 2024/1/solve.py
import argparse
from collections import Counter

def main(filename: str):

    with open(filename, "r") as file:
        lines: list[list[int]] = [list(map(int, line.strip().split())) for line in file.readlines()]

    l1: list[int] = []
    l2: list[int] = []

    for i in lines:
        l1.append(i[0])
        l2.append(i[1])

    l1.sort(); l2.sort()
    answer1: int = 0
    for i in range(len(l1)):
        answer1 += abs(l1[i] - l2[i])
    print(f"Answer 1: {answer1}")

    answer2: int = 0
    counter = Counter(l2)
    for i in range(len(l1)):
        answer2 += l1[i] * counter[l1[i]]
    print(f"Answer 2: {answer2}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("input_filename")
    args = parser.parse_args()

    main(args.input_filename)

File: 2024/1/solve.rs
use std::collections::HashMap;
use std::env;
use std::fs;
use std::str::FromStr;

fn main() {
    let args: Vec<String> = env::args().collect();
    let filename = &args[1];

    let lines: Vec<String> = fs::read_to_string(filename)
        .unwrap()
        .lines()
        .map(String::from)
        .collect();

    let mut v1: Vec<i32> = Vec::new();
    let mut v2: Vec<i32> = Vec::new();
    for line in lines {
        let values: Vec<&str> = line.split("   ").collect();
        v1.push(FromStr::from_str(values[0]).unwrap());
        v2.push(FromStr::from_str(values[1]).unwrap());
    }
    v1.sort();
    v2.sort();

    let mut answer1 = 0;
    for (i, _) in v1.iter().enumerate() {
        answer1 += (v1[i] - v2[i]).abs();
    }
    println!("Answer 1: {answer1}");

    let mut answer2 = 0;
    let mut counter: HashMap<i32, i32> = HashMap::new();
    let mut i = 0;
    while i < v2.len() {
        let mut count = 1;
        while i + 1 < v2.len() && v2[i] == v2[i + 1] {
            count += 1;
            i += 1
        }
        counter.insert(v2[i], count);
        i += 1;
    }
    for (_, value) in v1.iter().enumerate() {
        answer2 += *value * *counter.entry(*value).or_default();
    }
    println!("Answer 2: {answer2}");
}
